packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/jni/JNIDeallocator.kt

```diff
@@ -0,0 +1,73 @@
+package expo.modules.kotlin.jni
+
+import java.lang.ref.PhantomReference
+import java.lang.ref.ReferenceQueue
+import java.lang.ref.WeakReference
+
+interface Destructible {
+  fun deallocate()
+}
+
+object JNIDeallocator {
+  /**
+   * A [PhantomReference] queue managed by JVM
+   */
+  private val referenceQueue = ReferenceQueue<Destructible>()
+
+  /**
+   * A registry to keep all active [Destructible] objects and their [PhantomReference]s
+   */
+  private val destructorMap = mutableMapOf<PhantomReference<Destructible>, WeakReference<Destructible>>()
+
+  /**
+   * A thread that clears your registry when an object has been garbage collected
+   * to not store invalid references to every created object.
+   */
+  private val destructorThread = object : Thread("Expo JNI deallocator") {
+    override fun run() {
+      while (true) {
```

Comment:
does the thread keep running the loop? could we only run the loop when the referenceQueue is not empty? somehow using an locked object and wait for it.

Reply:
The `remove` method is blocking, so it will be paused most of the time.

Reply:
it's good to know. thanks @lukmccall!